feat(metrics): add NATS client metrics to prometheus_metrics_fmt

keivenchang opened the pull request:
#### Overview:

This PR adds comprehensive NATS metrics collection and Prometheus metrics management to the Dynamo runtime, including endpoint-grouped metrics and callback-based metric registration.

#### Details:

Simple Flow:

- Client hits /metrics
- Component callback pulls from SRV.STATS.<name>
- NATS callback pulls client metrics
- Both get copied into Prometheus structures
- Return formatted metrics to client

#### Where should the reviewer start?

- `lib/runtime/src/metrics.rs` - Core metrics registry implementation and callback system
- `lib/runtime/src/transports/nats.rs` - NATS metrics collection and registration
- `lib/runtime/src/component.rs` - Component metrics integration and automatic registration
- `lib/runtime/src/distributed.rs` - DistributedRuntime metrics management methods

#### Related Issues: (use one of the action words Closes / Fixes / Resolves / Relates to)

Relates to DIS-298

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Introduced a structured metrics registry system, allowing for registration and execution of custom metrics callbacks grouped by prefix.
  * Added integration of NATS client metrics, exposing internal NATS statistics as Prometheus metrics.
  * Enhanced metrics retrieval to execute registered callbacks before collecting Prometheus metrics.

* **Bug Fixes**
  * Standardized naming and logging for the system server (HTTP) across the runtime.

* **Tests**
  * Added tests to verify metrics registry callback functionality and NATS client metrics integration.

* **Refactor**
  * Improved organization of metrics registration and callback management for extensibility and clarity.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

keivenchang committed:
feat(metrics): add NATS metrics to prometheus_metrics_fmt

- Add NATS client metrics collection via callbacks in transports/nats.rs
- Execute callbacks before TextEncoder to include NATS metrics in output
- Add comprehensive test coverage for NATS client metrics
- Centralize NATS metric names in prometheus_names module
- Fixed runtime prefix hierarchy (bug previously)

keivenchang committed:
Update the name from register_metrics to register_metrics_callback

keivenchang committed:
refactor(metrics): switch to hierarchy key, skip duplicate registrations, reset on scrape errors; tidy tests and logs

keivenchang commented on lib/runtime/src/http_server.rs:
```diff
@@ -259,7 +259,9 @@ async fn metrics_handler(
+    state.drt().execute_metrics_callbacks("");
+
+    // Get all metrics from DistributedRuntime (top-level)
     match state.drt().prometheus_metrics_fmt() {
```
outdated

keivenchang commented on lib/runtime/src/transports/nats.rs:
```diff
@@ -489,6 +510,97 @@ impl NatsQueue {
+        let connects = stats.connects.load(Ordering::Relaxed);
+
+        // Get connection state
+        let connection_state = match format!("{:?}", client.connection_state()).as_str() {
+            "Connected" => 1,
+            "Connecting" => 2,
+            "Disconnected" => 0,
+            _ => 0, // Default to disconnected for unknown states
+        };
```
Fixed, in the next PR

keivenchang committed:
refactor: improve metrics callback API and NATS connection state handling

- Remove unnecessary _runtime parameter from MetricsRegistryEntry::add_callback
- Update add_callback calls to use new signature
- Replace brittle string formatting with proper enum matching for NATS State
- Update eprintln to tracing::error for consistent logging
- Fix test code to use updated callback signature

nnshah1 commented on lib/runtime/src/component.rs:
```diff
@@ -255,6 +260,57 @@ impl Component {
+        &self,
+    ) -> Result<crate::service::ComponentSystemStatusNatsMetrics> {
+        let component_metrics =
+            crate::service::ComponentSystemStatusNatsMetrics::from_component(self)?;
```
Question: why not import this type above instead of full reference here?

nnshah1 replied:
question: since this is a component system status object - would new() make sense here instead of from_component() ? i.e. that is it can only be created from a component.

nnshah1 commented on lib/runtime/src/component.rs:
```diff
@@ -255,6 +260,57 @@ impl Component {
+            crate::service::ComponentSystemStatusNatsMetrics::from_component(self)?;
+
+        // Create a callback that scrapes stats and updates metrics when called
+        let metrics_clone = component_metrics.clone();
```
question: why do we need this clone and the clone in the closure?

nnshah1 commented on lib/runtime/src/system_status_server.rs:
```diff
@@ -254,7 +255,10 @@ async fn metrics_handler(state: Arc<SystemStatusState>) -> impl IntoResponse {
 
-    // Get metrics from the registry
+    // Empty hierarchy means this is the top-level metrics at the DistributedRuntime level
+    state.drt().execute_metrics_callbacks("");
```
question: why the need for the hierarchy support if we are just doing top level - is it recursive?

keivenchang committed:
refactor(metrics): shorten imports; make callbacks return Result(()); store NATS client in metrics

keivenchang commented on lib/runtime/src/distributed.rs:
```diff
@@ -106,13 +106,21 @@ impl DistributedRuntime {
+            .nats_client()
+            .register_metrics_callback(&distributed_runtime)
+        {
+            tracing::warn!("Failed to register NATS client metrics: {}", e);
```
No longer valid (refactored).

keivenchang commented on lib/runtime/src/distributed.rs:
```diff
@@ -248,6 +256,75 @@ impl DistributedRuntime {
+        if let Some(entry) = registries.get(hierarchy) {
+            entry.execute_callbacks(self as &dyn crate::metrics::MetricsRegistry)
+        } else {
+            Vec::new()
```
The caller expects this to make several callback functions, and expects a Vec<Result> of the callback results. So if there's no callback, then it's just an empty Vec.

keivenchang committed:
refactor(metrics): callbacks accept Vec hierarchies; tighten docs and tests; log callback errors

keivenchang committed:
Make the timing more generous for CI

grahamking commented on lib/runtime/src/component.rs:
```diff
@@ -255,6 +261,58 @@ impl Component {
+        let component_clone = self.clone();
+        let mut hierarchies = self.parent_hierarchy();
+        hierarchies.push(self.hierarchy());
+        assert_eq!(
```
There's a `debug_assert` if you want, that is disabled in `cargo build --release` builds. That avoids doing the `cloned()` call in production builds.

keivenchang replied:
Got it thanks! Updated.

keivenchang commented on lib/runtime/src/component/namespace.rs:
```diff
@@ -84,7 +84,15 @@ impl MetricsRegistry for Namespace {
+                .filter(|name| !name.is_empty())
+                .collect();
+        names.reverse();
+        names.insert(0, String::new());
```
Done

keivenchang committed:
refactor: improve metrics collection with channel-based sync and better error handling

grahamking commented on lib/runtime/src/distributed.rs:
```diff
@@ -227,6 +244,76 @@ impl DistributedRuntime {
+            tracing::warn!(
+                hierarchy = ?hierarchy,
+                metric_name = ?metric_name,
+                "Metric already exists in registry; skipping registration"
```
Yay for structured logging!

grahamking commented on lib/runtime/src/distributed.rs:
```diff
@@ -227,6 +244,76 @@ impl DistributedRuntime {
+    /// Add a callback function to metrics registries for the given hierarchies
+    pub fn register_metrics_callback<F>(&self, hierarchies: Vec<String>, callback: F)
+    where
+        F: Fn() -> anyhow::Result<()> + Send + Sync + 'static,
```
I think you defined a type earlier for this.

grahamking commented on lib/runtime/src/distributed.rs:
```diff
@@ -227,6 +244,76 @@ impl DistributedRuntime {
+
+    /// Get all registered hierarchy keys. Private because it is only used for testing.
+    fn get_registered_hierarchies(&self) -> Vec<String> {
+        let registries = self.hierarchy_to_metricsregistry.lock().unwrap();
```
If you call `get_registered_hierarchies` a lot, a RWLock would be faster. Not sure if that's necessary.

I would expect all the metrics work to happen from a single thread, where the lock wouldn't be necessary.

grahamking commented on lib/runtime/src/distributed.rs:
```diff
@@ -227,6 +244,76 @@ impl DistributedRuntime {
+    /// Execute all callbacks for a given hierarchy key and return their results
+    pub fn execute_metrics_callbacks(&self, hierarchy: &str) -> Vec<anyhow::Result<()>> {
+        let registries = self.hierarchy_to_metricsregistry.lock().unwrap();
+        if let Some(entry) = registries.get(hierarchy) {
```
Could you release the lock before calling the callback?

keivenchang commented on lib/runtime/src/transports/nats.rs:
```diff
@@ -504,6 +507,99 @@ impl NatsQueue {
+        let connects = drt.create_intgauge(
+            nats_metrics::CONNECTS,
+            "Total number of connections established by NATS client",
+            &[],
```
I am copying from NATS server's own Statistics data structure (which are counters by nature), into our own Prometheus data structures. NATS Counters => Dynamo gauges. Make sense?

keivenchang commented on lib/runtime/src/metrics/prometheus_names.rs:
```diff
@@ -0,0 +1,128 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
```
Ahhhh the SPDX short-form license headers (when we have full text somewhere else). Got it, thanks!

keivenchang commented on lib/runtime/src/metrics/prometheus_names.rs:
```diff
@@ -0,0 +1,128 @@
+    nats::AVG_PROCESSING_MS,   // maps to: average_processing_time (nanoseconds)
+    nats::TOTAL_ERRORS,        // maps to: num_errors
+    nats::TOTAL_REQUESTS,      // maps to: num_requests
+    nats::TOTAL_PROCESSING_MS, // maps to: processing_time (nanoseconds)
```
Good catch & fixed, thanks! I originally started with ns, the type that was returned by NatsStatsMetrics (directly from $SRV.STATS.<name>) but updated it to be ms, but forgot to update the comment.

keivenchang commented on lib/runtime/src/service.rs:
```diff
@@ -79,20 +134,21 @@ impl EndpointInfo {
-    pub average_processing_time: f64,
+pub struct NatsStatsMetrics {
+    // Standard NATS Stats Service API fields from $SRV.STATS.<service_name> requests
+    pub average_processing_time: f64, // in nanoseconds according to nats-io
```
Old tests from several months ago indicated that these were floating point, which is wrong. Updated.

<img width="544" height="410" alt="Screenshot 2025-08-14 at 10 45 26 AM" src="https://github.com/user-attachments/assets/a22e71df-4521-43a0-add1-55f186340e8a" />

keivenchang commented on lib/runtime/src/service.rs:
```diff
@@ -244,3 +305,125 @@ mod tests {
+    /// Total requests across all endpoints (maps to: num_requests)
+    pub total_requests: prometheus::IntGauge,
+    /// Total processing time in milliseconds (maps to: processing_time)
+    pub total_processing_ms: prometheus::IntGauge,
```
Let me add a comment that these gauges are _COPIES_ of the total, which is in the NATS code.

keivenchang reviewed (commented):
Thanks for the comments and suggestions, learned quite a bit about Rust, really enjoyed learning new things! A new PR is coming up.

keivenchang committed:
refactor: improve metrics callback system with Arc and RwLock

- Use RuntimeCallback type alias with Arc for explicit sharing
- Replace Mutex with RwLock for better concurrent reads
- Minimize lock duration when executing callbacks
- Add tests with counter verification
- Document that metrics are snapshots, not live references

keivenchang committed:
fix: resolve clippy warnings and restore labels functionality

- Remove unused MetricsRegistry imports from multiple files
- Fix type mismatch in component retrieval logic (and_then -> map)
- Restore labels logic in metrics component
- Clean up unused code and improve code quality
- All tests passing except one pre-existing logging test issue

keivenchang committed:
fix: restore examples functionality by removing add_labels calls

- Remove add_labels method calls from all example files since method doesn't exist
- Clean up unused MetricsRegistry imports from example files
- Fix syntax errors in method chaining after removing add_labels
- All examples now compile and run successfully
- Examples are included in workspace for type inference during development

keivenchang committed:
fix: synchronize Python bindings with updated Rust runtime

- Remove add_labels method from Python bindings to match Rust API changes
- Python bindings now properly reflect the current Rust runtime interface
- Ensures consistency between Rust core and Python wrapper APIs

keivenchang committed:
style: apply code formatting to examples

- Format import statements for better readability
- Remove unnecessary blank lines in method chains
- Apply consistent code style across all example files
- Improve code formatting and organization
